samza-api/src/main/java/org/apache/samza/system/IncomingMessageEnvelope.java

```diff
@@ -71,6 +74,7 @@ public IncomingMessageEnvelope(SystemStreamPartition systemStreamPartition, Stri
     this.message = message;
     this.size = size;
     this.arrivalTime = Instant.now().toEpochMilli();
+    this.hashCodeForKeyHashComputation = key != null ? key.hashCode() : offset != null ? offset.hashCode() : hashCode();
```

Comment:
Is there a benefit to caching and storing it, rather than computing and exposing it via a function?

Reply:
I did consider this originally but if there are multiple calls to getSystemStreamPartitionKeyHash then it would be worth having this cached.